sql/expression/function/time.go

```diff
@@ -828,3 +904,111 @@ func weekFuncLogic(t time.Time) (interface{}, error) {
 	_, wk := t.ISOWeek()
 	return wk, nil
 }
+
+// TimeDiff subtracts the second argument from the first expressed as a time value.
+type TimeDiff struct {
+	expression.BinaryExpression
+}
+
+// NewTimeDiff creates a new NewTimeDiff expression.
+func NewTimeDiff(e1, e2 sql.Expression) sql.Expression {
+	return &TimeDiff{
+		expression.BinaryExpression{
+			Left:  e1,
+			Right: e2,
+		},
+	}
+}
+
+// Type implements the Expression interface.
+func (td *TimeDiff) Type() sql.Type { return sql.Time }
+
+// IsNullable implements the Expression interface.
+func (td *TimeDiff) IsNullable() bool { return false }
+
+func (td *TimeDiff) String() string {
+	return fmt.Sprintf("timediff(%s, %s)", td.Left, td.Right)
+}
+
+// WithChildren implements the Expression interface.
+func (td *TimeDiff) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(td, len(children), 2)
+	}
+	return NewTimeDiff(children[0], children[1]), nil
+}
+
+// Eval implements the Expression interface.
+func (td *TimeDiff) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	// check for valid types
+	if (td.Left.Type() == sql.Time || td.Left.Type() == sql.Timestamp ||
+		td.Left.Type() == sql.Datetime) && (td.Right.Type() == sql.Time ||
+		td.Right.Type() == sql.Timestamp || td.Right.Type() == sql.Datetime) {
+
+		// handle type mismatch
+		if td.Left.Type() != td.Right.Type() {
```

Comment:
`td.Left.Type() == sql.Timestamp` and `td.Right.Type() == sql.Datetime` will fail this check, but it's completely valid. `sql.Timestamp` is essentially a restricted version of `sql.Datetime`, similar to `int32` vs `int64`. You can compare the promoted types instead.

Reply:
I dont understand this... so youre saying `td.Left.Type() == sql.Timestamp` with `td.Right.Type() == sql.Datetime` should pass this check? What does `compare the promoted types instead` mean?